src/panels/lovelace/cards/energy/hui-energy-sankey-card.ts

```diff
@@ -0,0 +1,432 @@
+    }
+
+    // Group devices by areas and floors
+    const areas: Record<string, { value: number; devices: Node[] }> = {
+      no_area: {
+        value: 0,
+        devices: [],
+      },
+    };
+    const floors: Record<string, { value: number; areas: string[] }> = {
+      no_floor: {
+        value: 0,
+        areas: ["no_area"],
+      },
+    };
+    let untrackedConsumption = homeNode.value;
+    const computedStyle = getComputedStyle(this);
+    prefs.device_consumption.forEach((device, idx) => {
+      const entity = this.hass.entities[device.stat_consumption];
+      const value =
+        device.stat_consumption in this._data!.stats
+          ? calculateStatisticSumGrowth(
+              this._data!.stats[device.stat_consumption]
+            ) || 0
+          : 0;
+      if (value <= 0) {
+        return;
+      }
+      untrackedConsumption -= value;
+      const deviceNode: Node = {
+        id: device.stat_consumption,
+        label:
+          device.name ||
+          getStatisticLabel(
+            this.hass,
+            device.stat_consumption,
+            this._data!.statsMetadata[device.stat_consumption]
+          ),
+        value,
+        tooltip: `${formatNumber(value, this.hass.locale)} kWh`,
+        color: getGraphColorByIndex(idx, computedStyle),
+        index: 4,
+      };
+
+      const deviceArea = entity?.area_id;
+      if (deviceArea && this.hass.areas[deviceArea]) {
+        const area = this.hass.areas[deviceArea];
+
+        if (area.area_id in areas) {
+          areas[area.area_id].value += deviceNode.value;
+          areas[area.area_id].devices.push(deviceNode);
+        } else {
+          areas[area.area_id] = {
+            value: deviceNode.value,
+            devices: [deviceNode],
+          };
+        }
+        // see if the area has a floor
+        if (area.floor_id && this.hass.floors[area.floor_id]) {
+          if (area.floor_id in floors) {
+            floors[area.floor_id].value += deviceNode.value;
+            if (!floors[area.floor_id].areas.includes(area.area_id)) {
+              floors[area.floor_id].areas.push(area.area_id);
+            }
+          } else {
+            floors[area.floor_id] = {
+              value: deviceNode.value,
+              areas: [area.area_id],
+            };
+          }
+        } else {
+          floors.no_floor.value += deviceNode.value;
+          if (!floors.no_floor.areas.includes(area.area_id)) {
+            floors.no_floor.areas.unshift(area.area_id);
+          }
+        }
+      } else {
+        areas.no_area.value += deviceNode.value;
+        areas.no_area.devices.push(deviceNode);
+      }
+    });
+
+    Object.keys(floors)
+      .sort(
+        (a, b) =>
+          (this.hass.floors[b]?.level ?? -Infinity) -
+          (this.hass.floors[a]?.level ?? -Infinity)
+      )
+      .forEach((floorId) => {
+        let floorNodeId = `floor_${floorId}`;
+        if (this.hass.floors[floorId]) {
```

Comment:
```suggestion
        if (floorId in this.hass.floors) {
```

Reply:
the check is because the floor id here could be the special `no_floor`

Reply:
Shouldnt we just check for that specifically? I think it makes the code more understandable.

Reply:
yeah, makes sense